analyzers/src/SonarAnalyzer.CFG/Roslyn/CfgAllPathValidator.cs

```diff
@@ -23,29 +23,47 @@ namespace SonarAnalyzer.CFG.Roslyn
     public abstract class CfgAllPathValidator
     {
         private readonly ControlFlowGraph cfg;
-        private readonly Dictionary<BasicBlock, bool> visitedStatus = new Dictionary<BasicBlock, bool>();
+        private readonly HashSet<BasicBlock> visitedStatus = [];
 
         protected abstract bool IsValid(BasicBlock block);
         protected abstract bool IsInvalid(BasicBlock block);
 
         protected CfgAllPathValidator(ControlFlowGraph cfg) =>
             this.cfg = cfg;
 
-        public bool CheckAllPaths() =>
-            IsBlockOrAllSuccessorsValid(cfg.EntryBlock);
-
-        private bool IsBlockOrAllSuccessorsValid(BasicBlock block)
-        {
-            var isValid = !IsInvalid(block) && (IsValid(block) || AreAllSuccessorsValid(block));
-            visitedStatus[block] = isValid;
-            return isValid;
-        }
-
-        private bool AreAllSuccessorsValid(BasicBlock block)
+        public bool CheckAllPaths()
         {
-            visitedStatus[block] = true; // protects from loops, don't fail the computation if hits itself
-            return block.SuccessorBlocks.Any()
-                   && block.SuccessorBlocks.All(x => x != cfg.ExitBlock && (visitedStatus.ContainsKey(x) ? visitedStatus[x] : IsBlockOrAllSuccessorsValid(x)));
+            var blocks = new Stack<BasicBlock>();
+            blocks.Push(cfg.EntryBlock);
+            while (blocks.Count > 0)
+            {
+                var block = blocks.Pop();
+                if (IsInvalid(block))
+                {
+                    return false;
+                }
+                visitedStatus.Add(block); // protects from loops, don't fail the computation if hits itself
+                if (IsValid(block))
+                {
+                    continue;
+                }
+                if (block.SuccessorBlocks.IsEmpty)
+                {
+                    return false;
+                }
+                foreach (var successorBlock in block.SuccessorBlocks)
+                {
+                    if (successorBlock == cfg.ExitBlock)
+                    {
+                        return false;
+                    }
```

Comment:
To reduce the nesting of the code, and to take all decisions on the same level: What if we would do this check after `pop`, near the `IsInvalid`?
```
if (successorBlock == cfg.ExitBlock || IsInvalid(block))
{
    return false;
}
```

Reply:
`successorBlock` is the loop variable and not in scope above.